### PR TITLE
AJ-863 and AJ-865 filter by column

### PR DESCRIPTION
**Pull request by calypsomatic**
[AJ-863](https://broadworkbench.atlassian.net/browse/AJ-863) and [AJ-865](https://broadworkbench.atlassian.net/browse/AJ-865)
This PR exposes the filter-by-column and filter-by-name entity query parameters added in [AJ-864](https://broadworkbench.atlassian.net/browse/AJ-864) and [AJ-862](https://broadworkbench.atlassian.net/browse/AJ-862)
A search input appears in the column menu (for entity service/GCP workspaces only), triggering a search for an exact match in that column.
image
Mixpanel event added under `workspace:data:filteredSearch`

[AJ-863]: https://broadworkbench.atlassian.net/browse/AJ-863?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[AJ-865]: https://broadworkbench.atlassian.net/browse/AJ-865?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[AJ-864]: https://broadworkbench.atlassian.net/browse/AJ-864?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[AJ-862]: https://broadworkbench.atlassian.net/browse/AJ-862?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

When a table is filtered, the column it is filtered on will indicate the filter term:
image

### CODE REVIEW

**Comment by davidangb** [comment]
Some UX feedback, which may drive code changes so I haven't reviewed the code in depth yet:
* entering a value into the filter input and then clicking the X to clear the input triggers a new request which results in an error. Could you check if there is a value in the field before sending a request?
* entering a value and hitting enter causes the table to reload in the background, but the menu modal stays open. If I then hit the escape key, it triggers a new request which results in an error, but the menu modal still stays open.
* after performing a filter and seeing my subset of rows, there are no breadcrumbs or indicators of what I filtered on.
* after performing a filter and seeing my subset of rows, how do I go back to seeing the unfiltered results? I can click on the table name again in the left nav, but perhaps there should be a "remove filter" function? If we showed some breadcrumb, we could put an "X" next to that breadcrumb?

File: src/components/data/DataTable.js [hunk 1]
```diff
@@ -346,6 +372,12 @@ const DataTable = props => {
               }
             }
 
+            const filterBreadCrumb = h(div, { style: { display: 'flex', overflow: 'hidden' } }, [h(TextCell, { style: { fontWeight: 400, ...Style.noWrapEllipsis } }, [`filtered by: ${filterColTerm}`]), h(Clickable, {
```
Comment (davidangb) [inline src/components/data/DataTable.js]: I suggest changing `h(TextCell` to `h(TooltipCell` to get a nice tooltip for long filter values, like:
Screenshot 3-14-2023 at 05 32 PM

File: src/components/data/data-utils.js [hunk 2]
```diff
@@ -1185,13 +1187,26 @@ export const ModalToolButton = ({ icon, text, disabled, ...props }) => {
...
+  const columnMenu = h(PopupTrigger, {
+    ref: popup,
     closeOnClick: true,
     side: 'bottom',
     content: h(Fragment, [
       h(MenuButton, { onClick: () => onSort({ field, direction: 'asc' }) }, ['Sort Ascending']),
       h(MenuButton, { onClick: () => onSort({ field, direction: 'desc' }) }, ['Sort Descending']),
+      renderSearch && h(ConfirmedSearchInput, {
+        'aria-label': 'Filter by column',
+        placeholder: 'Exact match search in column',
```
Comment (davidangb) [inline src/components/data/data-utils.js]: nit: "filter" or "search"? I like "Exact match filter" better than "Exact match search" but either way the `aria-label` and the `placeholder` should be consistent

File: src/components/data/DataTable.js [hunk 3]
```diff
@@ -346,6 +372,12 @@ const DataTable = props => {
...
+              'aria-label': 'Clear filter',
+              tooltip: 'Clear filter',
+              style: { alignSelf: 'flex-start', marginLeft: 'auto' },
+              onClick: () => setColumnFilter('')
+            }, [icon('times-circle', { color: colors.light(8), size: 16 })])])
```
Comment (davidangb) [inline src/components/data/DataTable.js:374]: real minor nit: when the `Style.noWrapEllipsis` does NOT kick in - when the filter value is short - the "x" icon is awfully close to the value:
Screenshot 3-14-2023 at 05 34 PM

File: src/components/data/DataTable.js [hunk 4]
```diff
@@ -222,13 +226,33 @@ const DataTable = props => {
     return entities.length && _.every(k => _.includes(k, selectedKeys), entityKeys)
   }
 
+  const searchByColumn = (field, v) => {
+    setActiveTextFilter('')
+    setColumnFilter(`${field}=${v.toString().trim()}`)
```
Comment (davidangb) [inline src/components/data/DataTable.js]: I don't feel very strongly about this, so feel free to ignore: you could store `columnFilter` into state as an object e.g. `{ filterColAttr, filterColTerm }`; this prevents potentially-fragile string parsing in `parseColumnFilter`. Keep it as a structured object until creating the `queryOptions` object to pass to ajax; at that point, turn it into a string.

File: src/components/data/DataTable.js (hunk 3 above)
Comment (davidangb) [reply src/components/data/DataTable.js:374]: Also … when clicking the "x" to remove the filter … it changes the sort for that column as well :(

**Review by davidangb — COMMENTED** [review]
comments inline!

File: src/components/data/DataTable.js (hunk 3 above)
Comment (calypsomatic) [reply src/components/data/DataTable.js:374]: I cannot tell you how long I've spent trying to put space between the 'x' and the term.  I don't know why nothing I try works.

Comment (davidangb) [reply src/components/data/DataTable.js:374]: eh, then I say don't worry about it, let's get this PR in and potentially/maybe work on that spacing again later.

File: src/components/data/DataTable.js [hunk 5]
```diff
@@ -346,6 +372,12 @@ const DataTable = props => {
               }
             }
 
+            const filterBreadCrumb = h(div, { style: { display: 'flex', overflow: 'hidden' } }, [h(TextCell, { style: { fontWeight: 400, ...Style.noWrapEllipsis } }, [`filtered by: ${filterColTerm}`]), h(Clickable, {
+              'aria-label': 'Clear filter',
+              tooltip: 'Clear filter',
+              style: { alignSelf: 'flex-start', marginLeft: 'auto' },
```
Comment (davidangb) [inline src/components/data/DataTable.js]: for the "x" icon spacing, try this:
```suggestion
              style: { alignSelf: 'flex-start', marginLeft: '0.3rem' },
```

File: src/components/data/data-utils.js [hunk 6]
```diff
@@ -1185,13 +1187,26 @@ export const ModalToolButton = ({ icon, text, disabled, ...props }) => {
...
+        onChange: e => {
+          if (!!e) {
+            searchByColumn(e)
+            popup.current.close()
+          }
+        },
+        onClick: e => { e.stopPropagation() }
+      }),
```
Comment (davidangb) [inline src/components/data/data-utils.js]: maybe give the search field just a tiny bit of margin? I know we don't want to lose too much horizontal space here.
```suggestion
      renderSearch && h(div, { style: { width: '98%' } }, [h(ConfirmedSearchInput, {
        'aria-label': 'Filter by column',
        placeholder: 'Exact match search in column',
        style: { marginLeft: '0.25rem' },
        onChange: e => {
          if (!!e) {
            searchByColumn(e)
            popup.current.close()
          }
        },
        onClick: e => { e.stopPropagation() }
      })]),
```

File: src/components/data/DataTable.js (hunk 1 above)
Comment (calypsomatic) [reply src/components/data/DataTable.js]: Done!  Looks great!

File: src/components/data/DataTable.js (hunk 5 above)
Comment (calypsomatic) [reply src/components/data/DataTable.js]: Aha!  That did it!  Thank you!

File: src/components/data/data-utils.js (hunk 6 above)
Comment (calypsomatic) [reply src/components/data/data-utils.js]: Done

**Review by davidangb — APPROVED** [review]
love it, thank you!

**Review by aaronkanzer — APPROVED** [review]

**Pull request merged**